Network Resource Provisioning

shreeji818 opened the pull request:
New workflow to provision network resources

shreeji818 committed:
Network Resource Provisioning Workflow

shreeji818 committed:
Create .DS_Store

New workflows to provision network resources

shreeji818 committed:
Merge branch 'master' into master

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml:
```diff
@@ -0,0 +1,182 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
+<workflow>
+    <workflowInfo>
+        <workflowID>Derovision_Network_Resource</workflowID>
```
A few copypaste looking typos in the meta data here. 
De**p**rovision

Conner-IBM replied:
```suggestion
        <workflowID>Deprovision_Network_Resource</workflowID>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml:
```diff
@@ -0,0 +1,182 @@
+    <variable name="IP_ADDRESS_ID" visibility="private">
+        <label>IP_ADDRESS_ID</label>
+        <abstract>Object ID of the IP Address resource</abstract>
+        <description>The object id of the allocated IP address to use during deprovisioning</description>
```
The abstract uses uppercase ID while the description uses lower. This seems to be repeated in a few variables. I think uppercase 'ID" may be correct. Either way we should be consistent with casing I think.
```suggestion
        <description>The object ID of the allocated IP address to use during deprovisioning</description>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml:
```diff
@@ -0,0 +1,182 @@
+<workflow>
+    <workflowInfo>
+        <workflowID>Derovision_Network_Resource</workflowID>
+        <workflowDescription>Derovision Network Resource</workflowDescription>
```
```suggestion
        <workflowDescription>Deprovision Network Resource</workflowDescription>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml:
```diff
@@ -0,0 +1,182 @@
+
+    <variable name="IP_ADDRESS_ID" visibility="private">
+        <label>IP_ADDRESS_ID</label>
+        <abstract>Object ID of the IP Address resource</abstract>
```
```suggestion
        <abstract>Object ID of the IP address resource</abstract>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml:
```diff
@@ -0,0 +1,182 @@
+    <variable name="PORT_ID" visibility="private">
+        <label>PORT_ID</label>
+        <abstract>Object ID of the port resource</abstract>
+        <description>The object id of the reserved Port address to use during deprovisioning</description>
```
```suggestion
        <description>The object ID of the reserved port address to use during deprovisioning</description>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml:
```diff
@@ -0,0 +1,182 @@
+
+    <variable name="PORT_ID" visibility="private">
+        <label>PORT_ID</label>
+        <abstract>Object ID of the port resource</abstract>
```
```suggestion
        <abstract>Object ID of the port address</abstract>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml:
```diff
@@ -0,0 +1,182 @@
+    <variable name="APPLID_ID" visibility="private">
+        <label>APPLID_ID</label>
+        <abstract>Object ID of the APPLID resource</abstract>
+        <description>The object id of the reserved APPLID resource to use during deprovisioning</description>
```
```suggestion
        <description>The object ID of the reserved APPLID resource to use during deprovisioning</description>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml:
```diff
@@ -0,0 +1,182 @@
+        <description>Return allocated network resources back to Config Assistant</description>
+        <step name="returnDVIPA" optional="false">
+            <title>Return DVIPA</title>
+            <description>Invoke RESTful API to return IP address back to pool.</description>
```
```suggestion
            <description>Invoke RESTful API to return IP address back to the pool.</description>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml:
```diff
@@ -0,0 +1,182 @@
+                    <description>Set state to Skipped if IP_ADDRESS_ID is not set</description>
+                    <extendStateExpression>
+                        <description>Skip if IP_ADDRESS_ID is null</description>
+                        <expression>${instance_IP_ADDRESS_ID} == null</expression>
```
This looks fine, but I think these two conditions could be condensed into one by using velocity's silent notation.
Something like
`<expression>"$!{instance_IP_ADDRESS_ID}" == ""</expression>`
Should cover both cases

shreeji818 replied:
I understand that this can be collapsed but showing both conditions makes it clear to user that we are handling both null and empty values.

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml:
```diff
@@ -0,0 +1,182 @@
+                <description>Check that previous step has completed</description>
+                <targetStateSet>
+                    <description>Set state to Skipped if PORT_ID is not set.</description>
+                    <extendStateExpression>
```
Same comment as above can be applied here to use a single state expression

shreeji818 replied:
Please see previous response

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml:
```diff
@@ -0,0 +1,182 @@
+            <description>Invoke RESTful API to return APPLID back to pool.</description>
+            <condition>
+                <expression>"1" == "1"</expression>
+                <description>Check that previous step has completed</description>
```
Are we really checking that the previous step has completed here? This looks like its skipping the step depending on the APPLID_ID. Also same statement about condensing expressions

shreeji818 replied:
I will fix the comment

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml:
```diff
@@ -0,0 +1,182 @@
+                </targetStateSet>
+            </condition>
+            <instructions substitution="false">
+                 Execute step to return the VTAM APPLID back to z/OS Communications Server.
```
VTAM?  I see VTAM is being used by other workflows in this commit, but this is the only time this phrase is used during deprovisioning. Should it be used in other locations as well, or omitted here?

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,252 @@
+    </variable>
+    <variable name="IP_ADDRESS_ID" visibility="private">
+        <label>IP_ADDRESS_ID</label>
+        <abstract>Object ID of the IP Address resource</abstract>
```
```suggestion
        <abstract>Object ID of the IP address</abstract>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,252 @@
+    <variable name="IP_ADDRESS_ID" visibility="private">
+        <label>IP_ADDRESS_ID</label>
+        <abstract>Object ID of the IP Address resource</abstract>
+        <description>The object id of the allocated IP address to use during deprovisioning</description>
```
```suggestion
        <description>The object ID of the allocated IP address to use during provisioning</description>
```
We are using these resources during provisioning right?

shreeji818 replied:
This field is used during deprovisioning. So, comment is correct.

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,252 @@
+    </variable>
+    <variable name="PORT" visibility="public">
+        <label>PORT</label>
+        <abstract>TCP/IP Port Number</abstract>
```
```suggestion
        <abstract>TCP/IP port number</abstract>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,252 @@
+    <variable name="PORT" visibility="public">
+        <label>PORT</label>
+        <abstract>TCP/IP Port Number</abstract>
+        <description>TCP/IP port number</description>
```
This is the port we will be acquiring from the network pool right? Description could use just a little more clarity

shreeji818 replied:
done.

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,252 @@
+    </variable>
+    <variable name="PORT_ID" visibility="private">
+        <label>PORT_ID</label>
+        <abstract>Object ID of the port resource</abstract>
```
```suggestion
        <abstract>Object ID of the port address</abstract>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,252 @@
+    <variable name="PORT_ID" visibility="private">
+        <label>PORT_ID</label>
+        <abstract>Object ID of the port resource</abstract>
+        <description>The object id of the reserved Port address to use during deprovisioning</description>
```
```suggestion
        <description>The object ID of the reserved Port Address to use during provisioning</description>
```

Conner-IBM replied:
Or maybe instead of these suggestions, use the descriptions you provided in the README for these variables

shreeji818 replied:
Done

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/README.md:
```diff
@@ -0,0 +1,45 @@
+2. Create a new template for the Network Resource Provisioning following the instructions in IBM Cloud Provisioning and Management.
+3. Associate the template to a tenant (user or group of users) who will be requesting network resources and create network resource pool.
+4. Work with network administrator to complete network resource pool in z/OSMF Network Configuration Assistance plugin identifying IP ranges, port ranges and SNA APPL ID ranges.
+5. Test the template. Thevworkflow provided in "Provision_Network_Resource.xml" will be driven to provision the requested network resource.
```
```suggestion
5. Test the template. The workflow provided in "Provision_Network_Resource.xml" will be driven to provision the requested network resource.
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/README.md:
```diff
@@ -0,0 +1,45 @@
+3. Associate the template to a tenant (user or group of users) who will be requesting network resources and create network resource pool.
+4. Work with network administrator to complete network resource pool in z/OSMF Network Configuration Assistance plugin identifying IP ranges, port ranges and SNA APPL ID ranges.
+5. Test the template. Thevworkflow provided in "Provision_Network_Resource.xml" will be driven to provision the requested network resource.
+6. Publish the template 
```
```suggestion
6. Publish the template.
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/README.md:
```diff
@@ -0,0 +1,45 @@
+
+Once provisioning request completes, following properties are available in the provisioned instance based on the type of network resource requested at the provisioning time
+
+* IP_ADDRESS - This property provides the provisioned dynamic VIPA  when OBTAIN_DVIPA property was set at the provisioning time
```
I think that the descriptions here could be used to better describe some of the properties in the workflow definition itself. This would also help users co-relate what they see in these instructions to the workflow definition

shreeji818 replied:
Purpose of readme file is to provide additional info that is not in workflow and explain purpose of the service and input/output from service. But, I will see if the property explanation from read me can be applied to variable description in WF.

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,252 @@
+    <variable name="APPLID_ID" visibility="private">
+        <label>APPLID_ID</label>
+        <abstract>Object ID of the APPLID resource</abstract>
+        <description>The object id of the reserved APPLID resource to use during deprovisioning</description>
```
```suggestion
        <description>The object ID of the reserved APPLID resource to use during provisioning</description>
```

shreeji818 replied:
description is correct.

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,252 @@
+    <variable name="APPLID" visibility="public">
+        <label>APPLID</label>
+        <abstract>VTAM APPLID</abstract>
+        <description>VTAM APPLID</description>
```
This description could use some additional information

shreeji818 replied:
VTAM APPLID is known terminology on z/OS for network requests.  Will add some minor clarification

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,252 @@
+    <variable name="JOB_NAME" visibility="public">
+        <label>JOB_NAME</label>
+        <abstract>Job name of server that will use allocated port. Specify * if Jobname is not available</abstract>
+        <description>Job name of server that will use allocated port. Specify * if Jobname is not available</description>
```
Is this a question of availability, or more of a question as to if the user knows a specific job name to use?

shreeji818 replied:
Both.. In some cases e.g. when application runs as unix process, the jobname is internally generated and not available/known to user. I will change "available" to "known" to avoid any confusion

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,252 @@
+                    </description>
+                    <extendStateExpression>
+                        <description>Skip if OBTAIN_PORT does not equal YES</description>
+                        <expression>${instance_OBTAIN_PORT} != "YES"</expression>
```
This is a general comment that applies to all these variables used in these sorts of expressions.
Is there any reason we are doing a string match on "YES" instead of having these variables as booleans? They seem to only be used for expression checks. 
If there is a reason for using strings, I have an additional suggestion, but I think these should probably be booleans.

shreeji818 replied:
It does not matter if there is string compare or boolean. "YES"/"NO" reads better on choice to obtain resource vs TRUE/FALSE

Conner-IBM replied:
Functionally,I agree, it doesn't matter in this application.
If the purpose of this workflow is to be an example I don't think using YES/NO and doing a string compare for variables that describe themselves as booleans is the best example to make.

shreeji818 replied:
It's more readable for choice than True/False boolean. "Obtain_Port" value "No" vs "False" is more intuitive.
If value is specified via property file, we can define variable as boolean. But in this case, string makes more sense. So, lets go with that.

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,252 @@
+        <abstract>Usage type that matches with usage type associated with port pool</abstract>
+        <description>Usage type that matches with usage type associated with port pool. Request will fail if matching usage type is not found</description>
+        <category>REST</category>
+        <string valueMustBeChoice="false" multiLine="false"/>
```
Are there any example standard choices that we could provide here?

shreeji818 replied:
We won't be able to provide choice here because it depends on how network admin has setup the port ranges and what usage_type is defined there

Conner-IBM reviewed (changes requested):
I think that there are some improvements that could be made to the logical expressions and the variables that are being used there. I've made inline comments at these locations.
I've also added a bunch of suggestions around some of the text to make things a bit more consistent and fix a few typos.

Additionally, why did you include the .DS_Store file to this PR? That is just a binary file that should not be included.

 Let me know if you have any questions or concerns about the comments, Thanks.

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update Deprovision_Network_Resource.xml

Changed description for step expression

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/Deprovision_Network_Resource.xml

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update Provision_Network_Resource.xml

Added description to match readme file

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/README.md

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/README.md

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update Deprovision_Network_Resource.xml

shreeji818 committed:
Update README.md

shreeji818 committed:
Update README.md

shreeji818 committed:
Update Provision_Network_Resource.xml

shreeji818 reviewed (commented):
Updated files as per suggestions.

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,261 @@
+            <description>Allocate a DVIPA </description>
+            <condition>
+                <expression>"1" == "1"</expression>
+                <description>Check that previous step has completed</description>
```
This expression description seems inaccurate

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,261 @@
+    <variable name="APPLID_ID" visibility="private">
+        <label>APPLID_ID</label>
+        <abstract>Object ID of the APPLID name</abstract>
+        <description>The object id of the reserved APPLID name to use during deprovisioning</description>
```
```suggestion
        <description>The object ID of the reserved APPLID name to use during deprovisioning</description>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,261 @@
+    <variable name="PORT_ID" visibility="private">
+        <label>PORT_ID</label>
+        <abstract>Object ID of the port address</abstract>
+        <description>The object id of the reserved Port address to use during deprovisioning</description>
```
```suggestion
        <description>The object ID of the reserved port address to use during deprovisioning</description>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,261 @@
+    <variable name="IP_ADDRESS_ID" visibility="private">
+        <label>IP_ADDRESS_ID</label>
+        <abstract>Object ID of the IP address</abstract>
+        <description>The object id of the allocated IP address to use during deprovisioning</description>
```
```suggestion
        <description>The object ID of the allocated IP address to use during deprovisioning</description>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,261 @@
+    <variable name="IP_ADDRESS" visibility="public">
+        <label>IP_ADDRESS</label>
+        <abstract>TCP/IP DVIPA IP Address</abstract>
+        <description>TCP/IP DVIPA IP Address that will be obtained from the IP pool defined in the z/OSMF
```
```suggestion
        <description>TCP/IP DVIPA IP address that will be obtained from the IP pool defined in the z/OSMF
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,261 @@
+    </variable>
+    <variable name="IP_ADDRESS" visibility="public">
+        <label>IP_ADDRESS</label>
+        <abstract>TCP/IP DVIPA IP Address</abstract>
```
```suggestion
        <abstract>TCP/IP DVIPA IP address</abstract>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,261 @@
+    </variable>
+    <variable name="OBTAIN_APPLID" visibility="public">
+        <label>OBTAIN_APPLID</label>
+        <abstract>Obtain APPLID Choice</abstract>
```
```suggestion
        <abstract>Obtain APPLID choice</abstract>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,261 @@
+    </variable>
+    <variable name="OBTAIN_PORT" visibility="public">
+        <label>OBTAIN_PORT</label>
+        <abstract>Obtain Port Choice</abstract>
```
```suggestion
        <abstract>Obtain Port choice</abstract>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,261 @@
+    </workflowInfo>
+    <variable name="OBTAIN_DVIPA" visibility="public">
+        <label>OBTAIN_DVIPA</label>
+        <abstract>Obtain DVIPA Choice</abstract>
```
```suggestion
        <abstract>Obtain DVIPA choice</abstract>
```

Conner-IBM commented on provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml:
```diff
@@ -0,0 +1,261 @@
+            <description>Allocate VTAM APPLID </description>
+            <condition>
+                <expression>"1" == "1"</expression>
+                <description>Check that previous step has completed</description>
```
This expression description seems inaccurate

Conner-IBM reviewed (commented):
A few more suggestions for casing consistency and a follow up on an existing discussion.
Also please do not include the .DS_Store file.

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update provisioning-workflows/zOS/NetworkResources/Provision_Network_Resource.xml

Co-authored-by: Conner-IBM <conner.ferguson@ibm.com>

shreeji818 committed:
Update Provision_Network_Resource.xml

shreeji818 committed:
Update Deprovision_Network_Resource.xml

shreeji818 committed:
Revert "Create .DS_Store"

This reverts commit 19336c7557230a90080762fbb5d4bbce775212ed.

shreeji818 committed:
Added Instructions step

Fixed additional typos and added instructional step to avoid workflow failure when no property value is set.

shreeji818 committed:
Update Provision_Network_Resource.xml

shreeji818 committed:
Update Provision_Network_Resource.xml

shreeji818 committed:
Update Deprovision_Network_Resource.xml

shreeji818 commented:
@Conner-IBM Will you please Merge the pull request if there are no more comments.

Conner-IBM reviewed (approved):
All review comments where responded to.
